cudax/samples/vector_add_async/vector_add.cu

```diff
@@ -0,0 +1,142 @@
+#include <cuda/experimental/stream.cuh>
+
+#include <cstdio>
+#include <vector>
+
+#include "async_copy.cuh"
+#include "async_memory_resource.cuh"
+#include "async_uninitialized_buffer.cuh"
+
+namespace cudax = cuda::experimental;
+
+/**
+ * CUDA Kernel Device code
+ *
+ * Computes the vector addition of A and B into C. The 3 vectors have the same
+ * number of elements numElements.
+ */
+__global__ void vectorAdd(cuda::std::span<const float> A, cuda::std::span<const float> B, cuda::std::span<float> C)
+{
+  int i = blockDim.x * blockIdx.x + threadIdx.x;
+
+  if (i < A.size())
+  {
+    C[i] = A[i] + B[i] + 0.0f;
+  }
+}
+
+/**
+ * Host main routine
+ */
+int main(void)
+try
+{
+  // A CUDA stream on which to execute the vector addition kernel
+  cudax::stream stream(cudax::devices[0]);
+
+  // Print the vector length to be used, and compute its size
+  int numElements = 50000;
+  std::printf("[Vector addition of %d elements]\n", numElements);
+
+  // Allocate the host vectors
+  std::vector<float> h_A(numElements); // input
+  std::vector<float> h_B(numElements); // input
+  std::vector<float> h_C(numElements); // output
+
+  // Initialize the host input vectors
+  for (int i = 0; i < numElements; ++i)
+  {
+    h_A[i] = rand() / (float) RAND_MAX;
+    h_B[i] = rand() / (float) RAND_MAX;
+  }
+
+  // This is a resource for allocating device memory
+  cudax::cuda_async_memory_resource mr(stream.device());
+
+  // Asynchronously allocate the device buffers on the stream
+  cudax::async_uninitialized_buffer d_A(float{}, mr, numElements, stream);
+  cudax::async_uninitialized_buffer d_B(float{}, mr, numElements, stream);
+  cudax::async_uninitialized_buffer d_C(float{}, mr, numElements, stream);
+
+  // Asynchronously copy the host input data to the device
+  cudax::async_copy(h_A, d_A, stream);
+  cudax::async_copy(h_B, d_B, stream);
+
+  // Define the kernel launch parameters
+  constexpr int threadsPerBlock = 256;
+  auto dims                     = cudax::distribute<threadsPerBlock>(numElements);
+
+  // Launch the vectorAdd kernel
+  std::printf(
+    "CUDA kernel launch with %d blocks of %d threads\n", dims.count(cudax::block, cudax::grid), threadsPerBlock);
+  cudax::launch(stream, dims, vectorAdd, d_A, d_B, d_C);
+
+  // Copy the result from device to host
+  cudax::async_copy(d_C, h_C, stream);
+
+  std::printf("waiting for the stream to finish\n");
+  stream.wait();
+
+  std::printf("veryfying the results\n");
+  // Verify that the result vector is correct
+  for (int i = 0; i < numElements; ++i)
+  {
+    if (fabs(h_A[i] + h_B[i] - h_C[i]) > 1e-5)
+    {
+      std::fprintf(stderr, "Result verification failed at element %d!\n", i);
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  std::printf("Test PASSED\n");
+
+  std::printf("Done\n");
+  return 0;
+}
+catch (const std::exception& e)
+{
+  std::printf("caught an exception: \"%s\"\n", e.what());
+}
+catch (...)
+{
+  std::printf("caught an unknown exception\n");
+}
```

Comment:
Should we consider returning a nonzero error code from `main` in case it catches an exception?

```suggestion
  std::printf("caught an exception: \"%s\"\n", e.what());
  return -1;
}
catch (...)
{
  std::printf("caught an unknown exception\n");
  return -2;
}
```